### PR TITLE
V1 sys prompt word pad text out

### PR DESCRIPTION
**Pull request by sessid-nv**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: examples/speechlm2/s2s_duplex_speech_decoder_train.py [hunk 1]
```diff
@@ -24,32 +24,41 @@
...
+    
     trainer = Trainer(**resolve_trainer_cfg(cfg.trainer))
+    
     log_dir = exp_manager(trainer, cfg.get("exp_manager", None))
     OmegaConf.save(cfg, log_dir / "exp_config.yaml")
 
     with trainer.init_module():
         model = DuplexS2SSpeechDecoderModel(OmegaConf.to_container(cfg, resolve=True))
 
+    use_word_pad = cfg.model.tokenizer.get("use_word_pad", None)
```
Comment (nikhilsrihari18) [inline examples/speechlm2/s2s_duplex_speech_decoder_train.py:44]: Wondering if there is a way this can be moved either into the DuplexS2SDataset class itself (like by passing it the entire cfg object) OR in the DuplexS2SSpeechDecoderModel class (like prepare_inputs method). This way the code to train, which is more of a user view, would remain simple and similar to standard practice

Comment (nikhilsrihari18) [reply examples/speechlm2/s2s_duplex_speech_decoder_train.py:44]: Not a major item, but just a suggestion

File: nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py [hunk 2]
```diff
@@ -155,13 +249,26 @@ def __init__(self, cfg: dict) -> None:
...
+            self.cfg.pretrained_llm, pretrained_weights=self.cfg.pretrained_weights
+        ).train()
+        
+        # Add word padding tokens and prepare to learn new embeddings just for them
+        if self.cfg.get("tokenizer", None) and self.cfg.tokenizer.get("use_word_pad", None):
+            add_word_pad_token_embeddings(
+                llm, 
+                self.tokenizer,
+                cfg.get("tokenizer.train_new_embed_only", None)
+            )
```
Comment (nikhilsrihari18) [inline nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py:262]: This only gets called if tokenizer is specified in self.cfg? Shouldnt it always be called?

File: nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py [hunk 3]
```diff
@@ -1194,9 +1351,17 @@ def test_step(self, *args, **kwargs):
 
     def _get_bos_embedding(self) -> torch.Tensor:
         """
-        Remove the audio codec embedding for the beginning of AR decoding.
+
         """
-        text_bos = torch.full((1,), fill_value=self.text_pad_id, device=self.device)
+        if self.system_prompt is not None:
+            text_bos = torch.tensor(
+                    self.system_prompt_ids,
```
Comment (nikhilsrihari18) [inline nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py:1358]: self.system_prompt_ids is a method. Shouldnt this be a method call?

File: nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py [hunk 4]
```diff
@@ -330,6 +444,15 @@ def text_pad_id(self) -> int:
...
 
+    @property
+    def system_prompt_ids(self) -> Tensor:
+        system_prompt = self.cfg.get("system_prompt", "")
+        return self.tokenizer.tokenizer.convert_tokens_to_ids(
+            ['<|begin_of_text|>', '<|start_header_id|>'] +
+            ['system', '<|end_header_id|>'] +
+            self.tokenizer.tokenizer.tokenize(system_prompt) +
+            ['<|eot_id|>']
+        )
```
Comment (nikhilsrihari18) [inline nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py:455]: These seem hardcoded to nano v1 model. We will need to introduce separate logic for nano v2 model

File: nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py [hunk 5]
```diff
@@ -530,7 +656,10 @@ def prepare_inputs(self, batch: dict):
         Similar to DuplexS2SModel.prepare_inputs, with following changes:
             (1) Add 'input_audio_tokens' and 'seq_mask' in return value for TransformerARSpeechDecoder
             (2) Remove audio codec embedding from 'input_embeds'
+            ...
         """
+        ignore_speech_gen = self.cfg.get("ignore_speech_gen", None)
```
Comment (nikhilsrihari18) [inline nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py:661]: Set it as a self variable in init?

File: nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py (hunk 3 above)
Comment (nikhilsrihari18) [reply nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py:1358]: Addressed in https://github.com/nikhilsrihari18/NeMo/pull/6. Pls review

File: nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py (hunk 4 above)
Comment (nikhilsrihari18) [reply nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py:455]: Addressed in https://github.com/nikhilsrihari18/NeMo/pull/6. Pls review

File: nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py (hunk 5 above)
Comment (nikhilsrihari18) [reply nemo/collections/speechlm2/models/duplex_s2s_speech_decoder_model.py:661]: Addressed in https://github.com/nikhilsrihari18/NeMo/pull/6. Pls review
